docs/DesignDocs/DeepCopyOutParameters.md

```diff
@@ -0,0 +1,266 @@
+shared memory between a host and an enclave. Given that this support may not be general across TEEs,
+we discuss two designs in this section.
+
+The following code snippet serves as a common example that both designs aim to support.
+- Caller
+  ```
+  Blob b;
+  // b->len and b->buf have not been initialized yet.
+  foo(enclave, &b);
+  // b->len and b->buf have been initialized.
+  // b->len == 10 and b->data equals to data.
+  free(b->buf);
+  ```
+
+- Callee
+  ```
+  // User-defined code
+  void foo(Blob *b)
+  {
+    b->len = 10;
+    b->buf = malloc(b->size);
+    memcpy(b->buf, data, b->size);
+  }
+  ```
+  Note that `b->buf` also needs to be freed on the callee side, which is left to the implementation
+  details.
+
+### Design with Shared Memory
+
+With the support of shared memory, the design is more straightforward.
+The core idea of the design is that after the callee initializes the `b`,
+it also allocates a host memory and copies the content `b` to the allocated
+memory. Note that the latter part can be done by the oeedger8r-generated code.
+
+- ECALL
+  - Host
+
+    The host invokes the oeedger8r-generated ECALL stub as usual.
+    ```
+    // Oeedger8r-generated code
+    void foo(oe_enclave* enclave, Blob *b)
+    {
+      ...
+      oe_call_enclave_function(...);
+      ...
+    }
+    ```
+
+  - Enclave
+
+    Upon the dispatching of the ECALL, the oeedger8r-generated
+    `ecall_foo` is invoked as usual. However, right after the
+    invocation of the user-defined `foo`, the design adds a piece
+    of code that allocates host memory and copies the content
+    of `b` to the allocated memory.
+
+    ```
+    // Oeedger8r-generated code
+    void ecall_foo()
+    {
+      ...
+      Blob b;
+      foo(&b);
+      ...
+      // Initialize the host memory
+      Blob *lhs = (Blob *) pargs->in;
+      lhs->len = b.len;
+      lhs->buf = oe_host_malloc(lhs->len);
+      if (lhs->buf && b.buf)
+        memcpy(lhs->buf, b.buf, lhs->len);
+      ...
+    }
+    ```
+
+- OCALL
+  - Enclave
+
+    Given that the shared memory is available (i.e., the enclave
+    can access host memory), the design of the OCALL adds the similar
+    piece of the code on the enclave side that allocates the memory
+    and copies the content in the host memory to the allocated memory.
+
+    ```
+    // Oeedger8r-generated code
+    void foo(Blob *b)
+    {
+      ...
+      oe_call_host_function(...);
+      ...
+      if (b)
+      {
+        Blob *rhs = (Blob *) (_output_buffer + _output_buffer_offset);
+        b->len = rhs->len;
+        b->buf = oe_malloc(b->len);
+        if (b->buf && rhs->buf)
+          memcpy(b->buf, rhs->buf, b->len);
+      }
+    }
+    ```
+
+  - Host
+
+    In the case of OCALL, the callee invokes the user-defined `foo` as usual.
+    ```
+    // Oeedger8r-generated code
+    void ocall_foo()
+    {
+      ...
+      Blob b;
+      foo(&b);
+      ...
```

Comment:
why is there no free(b->buf); ?

Reply:
See my previous comment regarding the special case.